Fixes #307 & #354

Anders-Hopkins opened the pull request:
Added the Citation Modal to the GagePage so that it can be used to select the citations for new chars/stats and for editing chars/stats. Also made it so that only a single newchar/newstat/char/stat/GageInfo can be edited at a time. Made edits to the Citation Modal when it is opened in the GagePage: added an ID column and removed the last two columns.

Anders-Hopkins committed:
Initial Start

Added button to open citation modal when new characteristic is added. Modal opens, does not load citations yet.

Anders-Hopkins committed:
Use Citation Modal to select citation for new characteristic

In this push, the citation modal is used to select the citation for new characteristics in the gage page. The associated regions column has been removed from the citation modal. This still needs to be added to characteristics being edited as well as new/edited statistics.

Anders-Hopkins committed:
Use Citation Modal to select citations for new/edited characteristics in the gage page

- Citation modal used to select the citation for new characteristics
- Citation modal used to selected the citation for characteristics being edited
- Disabled multiple chars/stats from being edited at the same time, beginning edits on another char/stat will cancel the existing edits of a char/stat. Not sure if this is something we want to keep or not

Anders-Hopkins committed:
Merge branch 'dev' into AH-307

Anders-Hopkins committed:
Citation Modal used to add/edit citations to both chars & stats in the GagePage

Also, made it so that only a single char or stat can be edited at once. Style changes still need to be made.

Anders-Hopkins committed:
Citation Modal edits, updated the single row editing function, added ID column

Last changes before pull request. The Citation Modal is used to select the citation for new chars/stats and editing chars/stats. There is a function in place to prevent multiple chars/stats from being edited at the same time. The Citation Modal had to be edited to prevent the last two columns from displaying when opened in the GagePage. Also, added an ID column to the Citation Modal when it is opened in the GagePage.

Anders-Hopkins committed:
Cancel other edits when Edit Gage Info button is clicked

Cancel other edits when Edit Gage Info button is clicked.

kjacobsen16 commented:
@Anders-Hopkins I tried to pull this down and test it with one of the test stations, but I'm getting errors trying to edit or add characteristics or stats

Anders-Hopkins committed:
Made requested changes

Fixed issues that occurred on load of gagepage. Added functions to detele temp files when closing gagepage. Set a 5 seconds timer on toast popups.

Anders-Hopkins committed:
Update gagepage component

Opps, forgot to save all the files before I pushed this branch.

kjacobsen16 commented on src/app/shared/components/managecitations/managecitations.component.html:
```diff
@@ -50,87 +50,148 @@
+						<th *ngIf="!inGagePage" class=citationEditHeader></th>
+						<th *ngIf="inGagePage" class="width-5">ID</th>
+						<th *ngIf="inGagePage" class="width-85">Citation</th>
+						<th *ngIf="inGagePage" class="width-10"></th>
```
It looks like some of these are repeats, i.e. the citationHeader and citationEditHeader. Those two should be in there regardless. You can conditionally set the class if you need to, but otherwise just adding the ID here and keeping the others should be fine.

Anders-Hopkins replied:
Yea, so the reason why there are so many repeats here and especially the repeated columns below is because there needs to be different styling depending on whether the Citation Modal is being opened in a GagePage or from the MainView. Is there a way to make the CSS class dependent upon a Boolean variable? For instance; ngIf="inGagePage=true" class = A, else class = B.

kjacobsen16 replied:
Yes, you can conditionally set the class by doing something like `[class.width-5]="inGagePage"`

kjacobsen16 replied:
Another thought is you could set a conditional class higher up, something like `[class.gageStatsTable]="inGagePage"`, then use the existing classes here (citationHeader, etc.) combined with that class in the scss file, e.g. `.gageStatsTable.citationHeader` (it might be different for scss v. css)

kjacobsen16 commented on src/app/shared/components/managecitations/managecitations.component.html:
```diff
@@ -50,87 +50,148 @@
+										class="icon-button" aria-hidden="true">
+										<i class="far fa-plus"></i>
+									</button>
+								</div>
```
This is the section that allows a user to select a citation. This happens when a user in NSS is in the add regression region modal and hits "Find Existing Citation". You should be able to use this here and then just handle the click slightly differently.

kjacobsen16 commented on src/app/shared/components/managecitations/managecitations.component.html:
```diff
@@ -50,87 +50,148 @@
+									</a>
+									<span *ngIf="c.isEditing"><textarea class='break' name="citationURL"
+											[(ngModel)]="c.citationURL" required></textarea></span>
+								</div>
```
Again, this section and on has a lot of repeats that can be treated more or less the same if it's opened from gage page or not.

kjacobsen16 reviewed (changes requested):
Overall, a big improvement! A couple things I think would be helpful:
1. Putting a button in the main GageStats page to open the manage citation modal, should be pretty easy since it's already set up (can be a separate issue if you want)
2. This is a little confusing, but gagestats and nss have different citation tables, so the citations calls in the gagestats portion should actually use the `gagestatsservices/citations` endpoint
3. Can you make it more clear that clicking on a citation will select it? It took me a while to try just clicking on it since the cursor didn't change or anything. NSS has the "+" icon that you click to select it.
4. I can't see the full text for the "Add Citation" button. You could use an icon instead (like a magnifying glass or a list icon), or we can try to make the modal a bit bigger again.
![image](https://user-images.githubusercontent.com/35497731/98588877-08044800-2292-11eb-9bfc-6fe9aee323df.png)

Anders-Hopkins committed:
Fixed requested changes except for one bug

- Added Manage Citation button to the GageStats page
- Make citations selectable when opened in the GagePage
- Swapped "Add Citation" with plus sign in GagePage
- Pulls Citations from GSS rather than NSS

There is still a bug with filter citations functions.

Anders-Hopkins committed:
Fixes the filter() issue

Added an if statement to skip the filter() function is the variable this.citations is non-existant.

kjacobsen16 reviewed (changes requested):
Sorry to keep coming back with edits! There are more requests that need to go to the gagestats url instead. Namely, adding, editing and deleting citations in the manage citations modal hits nss instead of gagestats in the gagestats page (honestly at some point we should just move it to a shared table but that's far off).

We might want to consider changing how our settings service is doing the post/gagestatspost, etc. stuff, and instead of using the base url in the settings service, send it through in the component that's calling the request. That would require changing every instance that's calling them, but should be pretty simple. Let me know if that doesn't make sense!

Anders-Hopkins committed:
Fixed requested edits

- Removed getEntitiesGageStats(), postEntityGageStats, putEntityGageStats and deleteEntityGageStats from the settings.services.ts
- Edits the getEntities equalivent of all these functions to accept a general url inorder to make requests to both NSS and GSS
- Went thru and changes 140-ish instances of getEntityGageStats to getEntity
- Made should everything worked after these changes, some general clean up required
- Noticed that getRegions and getUnitTypes are sometimes slow to load and can have an effect on the displaying of tables KEEP AN EYE ON THIS

Anders-Hopkins committed:
Add/Edit/Delete NSS/GSS citations

-Added a function to set the url for http request to either NSS citations or GSS citations depending on which page is being viewed (Manview = NSS,  Gagestats & Gagepage = GSS)

Anders-Hopkins committed:
Merge branch 'dev' into AH-307

kjacobsen16 reviewed (approved):
Looks good - I'm adding a new issue to update the list of citations in the gagestats manage citations modal when there's an update to a citation (post, put, or delete)
